diff --git a/accel/kvm/kvm-all.c b/accel/kvm/kvm-all.c
index xxxxxxx..xxxxxxx 100644
--- a/accel/kvm/kvm-all.c
+++ b/accel/kvm/kvm-all.c
@@ -2018,7 +2018,12 @@ int kvm_init(MachineState *ms)
     s->vmfd = -1;
     s->fd = qemu_open("/dev/kvm", O_RDWR);
     if (s->fd == -1) {
+#ifdef __LIMBO__
+         //no m specifier in android
+        fprintf(stderr, "Could not access KVM kernel module: /dev/kvm\n");
+#else
         fprintf(stderr, "Could not access KVM kernel module: %m\n");
+#endif
         ret = -errno;
         goto err;
     }
diff --git a/audio/audio.c b/audio/audio.c
index xxxxxxx..xxxxxxx 100644
--- a/audio/audio.c
+++ b/audio/audio.c
@@ -2075,7 +2075,11 @@ static void audio_init_per_direction_opts(
 
     if (!pdo->has_frequency) {
         pdo->has_frequency = true;
+#ifdef __LIMBO__
+        pdo->frequency = 22050;
+#else
         pdo->frequency = 44100;
+#endif
     }
     if (!pdo->has_channels) {
         pdo->has_channels = true;
diff --git a/configure b/configure
index xxxxxxx..xxxxxxx 100755
--- a/configure
+++ b/configure
@@ -2504,7 +2504,8 @@ if test -z "$cross_prefix" && check_define __ANDROID__; then
 # pkg-config probe
diff --git a/include/qemu/osdep.h b/include/qemu/osdep.h
index xxxxxxx..xxxxxxx 100644
--- a/include/qemu/osdep.h
+++ b/include/qemu/osdep.h
@@ -110,6 +110,10 @@
 #include <libfdt.h>
 #endif
 
+#ifdef __LIMBO__
+#define stderr stdout
+#endif
+
 /*
  * This is somewhat insane, but it's possible to use _FORTIFY_SOURCE=2
  * even if the headers needed to implement it are not available.
diff --git a/migration/rdma.c b/migration/rdma.c
index xxxxxxx..xxxxxxx 100644
--- a/migration/rdma.c
+++ b/migration/rdma.c
@@ -33,6 +33,10 @@
  * Author: Michael R. Hines <mrhines@us.ibm.com>
  */
 
+#ifdef __LIMBO__
+#undef AF_INET
+#endif
+
 #include "qemu/osdep.h"
 #include "qemu/atomic.h"
 #include "qemu-common.h"
diff --git a/util/memfd.c b/util/memfd.c
index xxxxxxx..xxxxxxx 100644
--- a/util/memfd.c
+++ b/util/memfd.c
@@ -31,7 +31,11 @@
 #include "qemu/memfd.h"
 
 #ifdef CONFIG_MEMFD
+#ifdef __LIMBO__
+/* Not available on older android */
+#include "qemu/memfd_compat.h"
+#else
 #include <linux/memfd.h>
 #include <asm-generic/mman-common.h>
+#endif
 #endif